### PR TITLE
Merge commit of C API development effort

### PR DESCRIPTION
**Pull request by hoytak**
This commit exposes a C API to enable the direct wrapping of Turi Create
functionality by other languages.  Internal operations and structures, such as
SFrames, SArrays, and several ML models can now be used directly through an ABI
stable C interface.  This feature is designed to permit the use of Turi
Create as an embedded framework and facilitate the development of ML
applications in other languages besides C++ and python.

To build the C-API, run ./build_capi.sh in the base directory.  This
will generate a single dynamic library plus a single C header file that
allows complete interfacing to other languages.  Optionally, a Mac OSX
framework may be generated as well.

Currently, supported models are limited to the full supervised learning
suite; other models will be added in the future.

All SFrame and SArray operations are supported by the API.

In addition to the C-API, numerous under-the-hood improvements have been
made to allow for this interfacing, including:
- Updates to the extensions registration mechanism by which models are
exposed both to python and the C API.
- Updates to supervised learning so that it may be exposed through the
C API.
- Numerous small improvements and fixes to bugs exposed by testing the
code base through the C API.

### CODE REVIEW

File: src/capi/TuriCreate.h [hunk 1]
```diff
@@ -0,0 +1,861 @@
...
+void tc_parameters_add_flex_dict(tc_parameters*, const char* name, const tc_flex_dict* value, tc_error** error);
+void tc_parameters_add_datetime(tc_parameters*, const char* name, const tc_datetime* dt, tc_error**);
+void tc_parameters_add_image(tc_parameters*, const char* name, const tc_flex_image*, tc_error** error);
+void tc_parameters_add_flexible_type(tc_parameters*, const char* name, const tc_flexible_type*, tc_error** error);
+void tc_parameters_add_sarray(tc_parameters*, const char* name, const tc_sarray*, tc_error** error);
+void tc_parameters_add_sframe(tc_parameters*, const char* name, const tc_sframe*, tc_error** error);
+void tc_parameters_add_parameters(tc_parameters*, const char* name, const tc_parameters*, tc_error** error);
+void tc_parameters_add_model(tc_parameters*, const char* name, const tc_model*, tc_error** error);
+
+bool tc_parameters_is_int64(const tc_parameters*, const char* name, tc_error** error);
```
Comment (TobyRoseman) [inline src/capi/TuriCreate.h]: Do you want to use ints here? There are previous places in this file where a bool would have been appropriate but an int was used instead?

File: src/capi/TuriCreate.h [hunk 2]
```diff
@@ -0,0 +1,861 @@
...
+tc_flex_dict* tc_parameters_retrieve_flex_dict(const tc_parameters*, const char* name, tc_error**);
+tc_datetime* tc_parameters_retrieve_datetime(const tc_parameters*, const char* name, tc_error**);
+tc_flex_image* tc_parameters_retrieve_image(const tc_parameters*, const char* name, tc_error**);
+tc_flexible_type* tc_parameters_retrieve_flexible_type(const tc_parameters*, const char* name, tc_error**);
+tc_sarray* tc_parameters_retrieve_sarray(const tc_parameters*, const char* name, tc_error**);
+tc_sframe* tc_parameters_retrieve_sframe(const tc_parameters*, const char* name, tc_error**);
+tc_parameters* tc_parameters_retrieve_parameters(const tc_parameters*, const char* name, tc_error**);
+tc_model* tc_parameters_retrieve_model(const tc_parameters*, const char* name, tc_error**);
+
+// delete the parameter container.
```
Comment (TobyRoseman) [inline src/capi/TuriCreate.h]: Remove

File: src/capi/default_server_initializer.cpp [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+ *
+ *  This function creates the server initializer for the C-API, which is what
+ *  determines which models and functions are registered to the unity sever,
+ *  which then determines what models are packaged in the framework and
+ *  available from the C-API.
+ *
+ *  In some cases, it may be desirable to have a custom server initializer, for
+ *  example if only a subset of the models are needed.  In this case, define the
+ *  macro CAPI_DISABLE_DEFAULT_SERVER_INITIALIZER and create a custom
+ *  implementation of that function to be compiled in.  In addition, the CMakeLists.txt
```
Comment (TobyRoseman) [inline src/capi/default_server_initializer.cpp]: I don't understand this: ` In addition, the CMakeLists.txt`

File: src/capi/impl/capi_error_handling.cpp [hunk 4]
```diff
@@ -0,0 +1,55 @@
+#include <capi/TuriCreate.h>
+#include <capi/impl/capi_error_handling.hpp>
+#include <capi/impl/capi_initialization_internal.hpp>
+#include <string>
+#include <export.hpp>
+
+// These
```
Comment (TobyRoseman) [inline src/capi/impl/capi_error_handling.cpp]: Remove or finish

File: src/capi/impl/capi_initialization.hpp [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ *  This function creates the server initializer for the C-API, which is what
+ *  determines which models and functions are registered to the unity sever,
+ *  which then determines what models are packaged in the framework and
+ *  available from the C-API.
+ *
+ *  In some cases, it may be desirable to have a custom server initializer, for
+ *  example if only a subset of the models are needed.  In this case, define the
+ *  macro CAPI_DISABLE_DEFAULT_SERVER_INITIALIZER and create a custom
+ *  implementation of that function to be compiled in.  In addition, the CMakeLists.txt
```
Comment (TobyRoseman) [inline src/capi/impl/capi_initialization.hpp]: `In addition, the CMakeLists.txt`?

File: src/capi/impl/capi_models.cpp [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+EXPORT const char* tc_model_name(const tc_model* model, tc_error **error) { 
+  ERROR_HANDLE_START();
+  turi::ensure_server_initialized();
+
+  return model->value->name().c_str();
+  ERROR_HANDLE_END(error, "");
+}
+
+// Will be changed to
```
Comment (TobyRoseman) [inline src/capi/impl/capi_models.cpp]: ?

File: src/unity/toolkits/nearest_neighbors/class_registrations.hpp [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+#include <unity/lib/toolkit_class_macros.hpp>
+#include <unity/lib/toolkit_class_specification.hpp>
+
+namespace turi {
+namespace nearest_neighbors {
+
+std::vector<turi::toolkit_class_specification> get_toolkit_class_registration();
+
+}// nearest_neighbors
+}// recsys
```
Comment (TobyRoseman) [inline src/unity/toolkits/nearest_neighbors/class_registrations.hpp]: `recsys` -> `turi`

File: src/unity/toolkits/supervised_learning/class_registrations.cpp [hunk 8]
```diff
@@ -0,0 +1,51 @@
+#include <unity/toolkits/supervised_learning/class_registrations.hpp>
```
Comment (TobyRoseman) [inline src/unity/toolkits/supervised_learning/class_registrations.cpp]: license header

File: src/unity/toolkits/supervised_learning/class_registrations.hpp [hunk 9]
```diff
@@ -0,0 +1,17 @@
+#ifndef TURI_SUPERVISED_CLASS_REGISTRATIONS
```
Comment (TobyRoseman) [inline src/unity/toolkits/supervised_learning/class_registrations.hpp]: license header

File: src/unity/toolkits/text/class_registrations.cpp [hunk 10]
```diff
@@ -0,0 +1,12 @@
+#include <unity/toolkits/text/alias.hpp>
```
Comment (TobyRoseman) [inline src/unity/toolkits/text/class_registrations.cpp]: license header

File: src/unity/toolkits/text/class_registrations.hpp [hunk 11]
```diff
@@ -0,0 +1,16 @@
+#ifndef TURI_TEXT_CLASS_REGISTRATIONS
```
Comment (TobyRoseman) [inline src/unity/toolkits/text/class_registrations.hpp]: license header

File: test/capi/capi_classifier_evaluation.cxx [hunk 12]
```diff
@@ -0,0 +1,87 @@
+#define BOOST_TEST_MODULE capi_classifier_evaluations
```
Comment (TobyRoseman) [inline test/capi/capi_classifier_evaluation.cxx]: license header

File: test/capi/capi_datetime.cxx [hunk 13]
```diff
@@ -0,0 +1,192 @@
+#define BOOST_TEST_MODULE capi_datetime
```
Comment (TobyRoseman) [inline test/capi/capi_datetime.cxx]: license header

File: src/capi/TuriCreate.h [hunk 14]
```diff
@@ -0,0 +1,861 @@
+#ifndef TURI_CAPI_H
```
Comment (TobyRoseman) [inline src/capi/TuriCreate.h]: license header

File: src/capi/default_server_initializer.cpp [hunk 15]
```diff
@@ -0,0 +1,22 @@
+#include <unity/server/unity_server_control.hpp>
```
Comment (TobyRoseman) [inline src/capi/default_server_initializer.cpp]: license header

File: build_python_egg.sh [hunk 16]
```diff
@@ -0,0 +1,124 @@
+#!/bin/bash -e
```
[the remark this answers is not in the document]
Comment (hoytak) [reply build_python_egg.sh]: done

File: src/capi/CMakeLists.txt [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+make_empty_library(TuriCreate_static
+  REQUIRES
+  unity _turi_capi_default_initializer _turi_capi_core)
+
+# Fix the current source directory of the framework stuff so the macro below
+# references everything correctly.
+set(__CAPI_SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR} CACHE INTERNAL "")
+
+# A macro that allows building a custom C-API library.
+# TODO: documentation
```
[the remark this answers is not in the document]
Comment (hoytak) [reply src/capi/CMakeLists.txt]: done

File: src/capi/TuriCreate.h [hunk 18]
```diff
@@ -0,0 +1,861 @@
...
+
+/******************************************************************************/
+/*                                                                            */
+/*    INITIALIZATION                                                          */
+/*                                                                            */
+/******************************************************************************/
+
+/** Initializing the framework. 
+ *
+ *  Call these before calling any non-setup function.
```
[the remark this answers is not in the document]
Comment (hoytak) [reply src/capi/TuriCreate.h]: Currently 1, more in future.

File: src/capi/TuriCreate.h [hunk 19]
```diff
@@ -0,0 +1,861 @@
...
+ *  Return object is a null-terminated c-style message string.
+ *
+ *  The char buffer returned is invalidated by calling tc_error_destroy.
+ */
+const char* tc_error_message(const tc_error* error);
+
+
+/** Destroys any types.
+ *
+ *  Must be called on the
```
[the remark this answers is not in the document]
Comment (hoytak) [reply src/capi/TuriCreate.h]: done

File: src/unity/toolkits/text/class_registrations.hpp (hunk 11 above)
Comment (hoytak) [reply src/unity/toolkits/text/class_registrations.hpp]: Done

File: test/capi/capi_classifier_evaluation.cxx (hunk 12 above)
Comment (hoytak) [reply test/capi/capi_classifier_evaluation.cxx]: Done

File: src/capi/TuriCreate.h (hunk 14 above)
Comment (hoytak) [reply src/capi/TuriCreate.h]: Done

**Pull request merged**